compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/types/BallerinaRecordTypeBuilder.java

```diff
@@ -0,0 +1,255 @@
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BVarSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.Symbols;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BField;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BRecordType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BType;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+import org.wso2.ballerinalang.compiler.util.Names;
+import org.wso2.ballerinalang.util.Flags;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+/**
+ * The implementation of the methods used to build the Record type descriptor in Types API.
+ *
+ * @since 2.0.0
+ */
+public class BallerinaRecordTypeBuilder implements TypeBuilder.RECORD {
+
+    private final TypesFactory typesFactory;
+    private final SymbolTable symTable;
+    private TypeSymbol restField;
+    private final List<RECORD_FIELD> recordFieldList = new ArrayList<>();
+    private final List<TypeReferenceTypeSymbol> typeInclusions = new ArrayList<>();
+
+    public BallerinaRecordTypeBuilder(CompilerContext context) {
+        typesFactory = TypesFactory.getInstance(context);
+        symTable = SymbolTable.getInstance(context);
+    }
+
+    @Override
+    public RECORD_FIELD fields() {
+        return new RecordFieldDataHolder();
+    }
+
+    @Override
+    public TypeBuilder.RECORD withFields(RECORD_FIELD... fields) {
+        recordFieldList.clear();
+        recordFieldList.addAll(Arrays.asList(fields));
+
+        return this;
+    }
+
+    @Override
+    public TypeBuilder.RECORD withRestField(TypeSymbol restType) {
+        restField = restType;
+
+        return this;
+    }
+
+    @Override
+    public TypeBuilder.RECORD withTypeInclusions(TypeReferenceTypeSymbol... typeInclusions) {
+        this.typeInclusions.clear();
+        this.typeInclusions.addAll(Arrays.asList(typeInclusions));
+
+        return this;
+    }
+
+    @Override
+    public RecordTypeSymbol build() {
+
+        BRecordTypeSymbol recordSymbol = Symbols.createRecordSymbol(0, Names.EMPTY, symTable.rootPkgSymbol.pkgID, null, symTable.rootPkgSymbol.owner,
+                symTable.builtinPos, symTable.rootPkgSymbol.origin);
+
+        BRecordType recordType = new BRecordType(recordSymbol);
+        recordSymbol.type = recordType;
+        recordType.typeInclusions = getTypeInclusions(typeInclusions);
+        if (restField == null) {
+            recordType.restFieldType = symTable.noType;
+        } else {
+            recordType.restFieldType = getBType(restField);
+        }
+
+        recordType.fields = getFields(recordFieldList, recordSymbol);
+        RecordTypeSymbol recordFieldSymbol = (RecordTypeSymbol) typesFactory.getTypeDescriptor(recordType);
+
+        typeInclusions.clear();
+        restField = null;
+        recordFieldList.clear();
```

Comment:
Are we doing this to make sure that we don't create another same-attributed type using the same builder?

Reply:
Discussed with the team. The same builder will be used to create another type